toplevel/toploop.ml

```diff
@@ -206,21 +218,42 @@ let loop ppf =
   run_hooks After_setup;
   load_ocamlinit ppf;
   while true do
-    let snap = Btype.snapshot () in
+    let snap = ref (Btype.snapshot ()) in
     try
       Lexing.flush_input lb;
       (* Reset the phrase buffer when we flush the lexing buffer. *)
       Buffer.reset phrase_buffer;
       Location.reset();
-      Warnings.reset_fatal ();
       first_line := true;
-      let phr = try !parse_toplevel_phrase lb with Exit -> raise PPerror in
-      let phr = preprocess_phrase ppf phr  in
-      Env.reset_cache_toplevel ();
-      ignore(execute_phrase true ppf phr)
+      let rec get_phrases phrs =
+        match !parse_toplevel_phrase lb with
+        | phr ->
+          if is_blank lb then
+            List.rev (phr :: phrs)
+          else
+            get_phrases (phr :: phrs)
+        | exception Exit -> raise PPerror
+        | exception e -> Location.report_exception ppf e; []
+      in
+      let phrs = get_phrases [] in
+      let saved_report_printer = !Location.report_printer in
+      let nonfirst_report_printer = fun () -> Location.batch_mode_printer in
+      let snap = ref (Btype.snapshot ()) in
+      let finally () = Location.report_printer := saved_report_printer in
+      Fun.protect ~finally begin fun () ->
+        let execute i phr =
+          if i > 0 then Location.report_printer := nonfirst_report_printer;
+          snap := Btype.snapshot ();
+          Warnings.reset_fatal ();
+          let phr = preprocess_phrase ppf phr in
+          Env.reset_cache_toplevel ();
+          ignore(execute_phrase true ppf phr)
+        in
+        List.iteri execute phrs
```

Comment:
Note: another way to write this logic which might be nicer would something like this (sketch):

```ocaml
let rec execute_phrases ~first_phrase = function
| [] -> ()
| phrase :: rest ->
    snap := Btype.snapshot ();
    Warnings.reset_fatal ();
    let phr = preprocess_phrase ppf phr in
    Env.reset_cache_toplevel ();
    ignore (execute_phrase true ppf phr);
    if not first_phrase then execute_phrases ~first_phrase:false rest
    else
      Misc.protect_refs [R Location.(report_printer, batch_printer)] (fun () ->
        execute_phrases ~first_phrase:false rest)
in
execute_phrases ~first_phrase:true phrs
```

Reply:
Thanks for the suggestion.